### PR TITLE
PC-1465: stop old epcs showing in epc select

### PR DESCRIPTION
**Pull request by Glenn-Clarke**
[Link to Jira ticket](https://beisdigital.atlassian.net/browse/PC-1465)

# Description

Filters the returned list of EPCs for E&W to only include the most recent EPC for each UPRN present in the list.

# Checklist

- [x] I have made any necessary updates to the documentation
- [x] I have checked there are no unnecessary IDE warnings in this PR
- [x] I have checked there are no unintentional line ending changes
- [x] I have added tests where applicable
- [x] If I have made any changes to the code, I have used the IDE auto-formatter on it
- [x] If I have made any changes to Python files (even if not changing any content strings), I have run `make extract-translations`
- [x] If I have made any changes to website flow, I have updated the [Flow Miro Board](https://miro.com/app/board/uXjVK4AgQ3E=/)
- [x] If I have made any changes to website flow, I have checked forward and back behaviour is still consistent

### CODE REVIEW

File: tests/test_frontdoor.py
Comment (samyou-softwire) [inline tests/test_frontdoor.py:2149]: perhaps better to use the defined consts for the keys here (& maybe page names too?)

File: help_to_heat/utils.py
Comment (samyou-softwire) [inline help_to_heat/utils.py:260]: does this work fine for epcs with no uprn?

Comment (samyou-softwire) [inline help_to_heat/utils.py:260]: is it necessary to wrap `group` in a list here? would assume its already iterable

Comment (samyou-softwire) [inline help_to_heat/utils.py:260]: could the variable name be improved here?

File: help_to_heat/utils.py
Comment (samyou-softwire) [inline help_to_heat/utils.py:257]: may be better to use [`sorted`](https://www.w3schools.com/python/ref_func_sorted.asp) here to avoid mutating the list we pass in?

File: tests/test_frontdoor.py
Comment (samyou-softwire) [inline tests/test_frontdoor.py:2146]: for what its worth it may be more in the spirit of these tests to check what text appears on the page rather than the underlying answers. maybe both?

**Review by samyou-softwire — CHANGES_REQUESTED** [review]

File: tests/test_frontdoor.py
Comment (Glenn-Clarke) [reply tests/test_frontdoor.py:2146]: I think you're right that the page's appearing text should be checked, but the underlying data is critical to determine whether it's the correct one of the two identical addresses showing.

File: help_to_heat/utils.py
Comment (Glenn-Clarke) [reply help_to_heat/utils.py:260]: I didn't realise this was an issue, I'll add a case where this ignores no UPRN cases, since as discussed it hasn't been seen yet in practice, but is a niche possibility.

File: help_to_heat/utils.py
Comment (samyou-softwire) [inline help_to_heat/utils.py:257]: seems like this lambda could be replaced with an [`attrgetter`](https://docs.python.org/3/howto/sorting.html#operator-module-functions-and-partial-function-evaluation)

Comment (Glenn-Clarke) [reply help_to_heat/utils.py:257]: Oh nice! I hadn't seen this before, I'll use it below as well

File: help_to_heat/utils.py
Comment (samyou-softwire) [reply help_to_heat/utils.py:260]: for what its worth I've added this caveat to our docs here: https://softwiretech.atlassian.net/wiki/spaces/Support/pages/20562509848/DESNZ+GBIS+Help+to+Heat+EPCs#EPC-UPRNs

File: help_to_heat/utils.py
Comment (samyou-softwire) [inline help_to_heat/utils.py:259]: I suppose it's not clear at this time whether a missing UPRN will come through as an empty string or be vacant entirely; probably safer to use the `.get()` method rather than indexing to provide a sensible default & avoid an error being thrown (probably `""` to align with the below line)

probably safer to use `.get()` in all cases here

File: tests/test_frontdoor.py
Comment (samyou-softwire) [inline tests/test_frontdoor.py:2152]: we would expect two addresses to be present? feels like this is what the below line indicates

**Review by samyou-softwire — CHANGES_REQUESTED** [review]

File: tests/test_frontdoor.py
Comment (Glenn-Clarke) [reply tests/test_frontdoor.py:2152]: We expect only one instance of this visible address value. 

The line below is checking the data to ensure that neither is the excluded one on an LMK level, since we can't tell which of the above address line is shown at the current screen.

Was trying to keep the test focused on this address, given the focus of this test is about pinpointing that the newest EPC is found specifically.

**Review by samyou-softwire — APPROVED** [review]

File: help_to_heat/utils.py
Comment (jdgage) [inline help_to_heat/utils.py:260]: Think we should document here why we do this (i.e. what's in our docs) and our justification for including them all in this case.

File: tests/test_frontdoor.py
Comment (jdgage) [reply tests/test_frontdoor.py:2152]: I don't think I follow this - why do we expect the other address?

Comment (Glenn-Clarke) [reply tests/test_frontdoor.py:2152]: The other one should also be expected, because it's the newest of its UPRN. So really both should definitely appear if the grouping is working as expected

I'm getting a bit lost reading this whole thing back, so to reset, as the test is written now:
- The sample data contains 3 EPCs. Two for one address, one for another.
- We expect this page to show two radio button entries, one for each of the two UPRNs.
- We expect neither of the two radios has the LMK matching the old EPC for the property that has two EPCs.

This should show that we're getting the newest EPC for each UPRN.

Comment (jdgage) [reply tests/test_frontdoor.py:2152]: Ok - would it be clearer what we're testing if there was only the two for the one address in the test data? Feels like we've introduced some noise to the test without needing to: ideally unit tests should test just one thing, so there's not a risk of them failing when something else has gone wrong (making it harder to debug that thing).

**Review by jdgage — APPROVED** [review]
Approving so as not to hold up releasing the fix while I quibble about tests

File: tests/test_frontdoor.py
Comment (Glenn-Clarke) [reply tests/test_frontdoor.py:2152]: I've noticed 11 Acacia never appears, at least by address. I'm running the tests quickly to see if taking it out breaks anything.

**Pull request merged**
